spring-cloud-app-broker-logging/src/main/java/org/springframework/cloud/appbroker/logging/streaming/ApplicationLogStreamPublisher.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2002-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.appbroker.logging.streaming;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+
+import org.springframework.cloud.appbroker.logging.streaming.events.ServiceInstanceLogEvent;
+import org.springframework.cloud.appbroker.logging.streaming.events.ServiceInstanceLoggingEvent;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.context.ApplicationListener;
+
+public class ApplicationLogStreamPublisher implements ApplicationListener<ServiceInstanceLoggingEvent> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ApplicationLogStreamPublisher.class);
+
+	private final Map<String, Registration> registry = new HashMap<>();
+
+	private final LogStreamPublisher<org.cloudfoundry.dropsonde.events.Envelope> logStreamPublisher;
+
+	private final ApplicationEventPublisher publisher;
+
+	public ApplicationLogStreamPublisher(
+		LogStreamPublisher<org.cloudfoundry.dropsonde.events.Envelope> logStreamPublisher,
+		ApplicationEventPublisher publisher) {
+		this.logStreamPublisher = logStreamPublisher;
+		this.publisher = publisher;
+	}
+
+	@Override
+	public void onApplicationEvent(ServiceInstanceLoggingEvent event) {
+		final String serviceInstanceId = event.getServiceInstanceId();
+		switch (event.getOperation()) {
+			case START:
+				LOG.debug("Received event to begin listening to logs for {}", serviceInstanceId);
+				this.startPublishing(serviceInstanceId);
+				return;
+			case STOP:
+				LOG.debug("Received event to stop listening to logs for {}", serviceInstanceId);
+				this.stopPublishing(serviceInstanceId);
+				return;
+
+			default:
+				throw new IllegalArgumentException("Unknown operation: " + event.getOperation());
+		}
+	}
+
+	private void startPublishing(String serviceInstanceId) {
+		synchronized (registry) {
```

Comment:
I think this could be handled with `ConcurrentMap.compute` rather than synchronising on the whole map instance. We're only ever updating the value for a single key here, so why lock access to the whole map?

Reply:
Agreed! Good catch!